spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/SpanSubscriber.java

```diff
@@ -44,35 +46,40 @@
 
 	private final Span span;
 
-	private final Span rootSpan;
+	private final TraceContext parent;
 
 	private final Subscriber<? super T> subscriber;
 
 	private final Context context;
 
 	private final Tracer tracer;
 
+	private final CurrentTraceContext currentTraceContext;
+
 	private Subscription s;
 
 	SpanSubscriber(Subscriber<? super T> subscriber, Context ctx, Tracing tracing,
 			String name) {
 		this.subscriber = subscriber;
 		this.tracer = tracing.tracer();
-		Span root = ctx.getOrDefault(Span.class, this.tracer.currentSpan());
+		this.currentTraceContext = tracing.currentTraceContext();
+		TraceContext parent = ctx.getOrDefault(TraceContext.class, null);
+		if (parent == null) {
+			parent = currentTraceContext.get();
+		}
 		if (log.isTraceEnabled()) {
-			log.trace("Span from context [{}]", root);
+			log.trace("Span from context [{}]", parent);
 		}
-		this.rootSpan = root;
+		this.parent = parent;
 		if (log.isTraceEnabled()) {
-			log.trace("Stored context root span [{}]", this.rootSpan);
+			log.trace("Stored context parent span [{}]", this.parent);
 		}
-		this.span = root != null ? this.tracer
-				.nextSpan(TraceContextOrSamplingFlags.create(root.context())).name(name)
-				: this.tracer.nextSpan().name(name);
```

Comment:
nextSpan makes a get call to the current context, which is not needed when we know we are in a new root scenario.